File: projects/engineering.purdue.edu/scriptingwithobjects/swocode/chap15/AcceptAndForkServer.pl
#!/usr/bin/perl -w
use strict;

### ServerSocket.pl

use IO::Socket;       # for IO::Socket::INET                         #(A)
use Net::hostent;     # for 1-arg version of gethostbyaddr           #(B)

sub prefix {                                                         #(C)
    my $now = localtime;                                             #(D)
    "[$now] [$$]" . shift( @_ );                                     #(E)
}

$SIG{__WARN__} = sub { warn prefix shift };                          #(F)
$SIG{__DIE__} = sub { die prefix shift };                            #(G)
$SIG{CLD} = $SIG{CHLD} = sub { wait; };                              #(H)

my $server_sock = IO::Socket::INET->new(LocalPort => 9000,           #(I)
                                        Listen => 5,                 #(J)
                                        Reuse => 1,                  #(K)
                                        Proto => 'tcp',              #(L)
                                        );
die "No Server Socket: $@\n" unless $server_sock;                    #(M)

while (1) {                                                          #(N)
    my $client_sock = $server_sock->accept();                        #(O)
    next unless defined $client_sock;                                #(P)
    my $pid = fork();                                                #(Q)
    die "Cannot fork: $!" unless defined( $pid );                    #(R)
    # child process:
    if ($pid == 0) {                                                 #(S)
        $server_sock->close;                                         #(T)
        my $hostinfo = gethostbyaddr($client_sock->peeraddr);        #(U)
        my $hostname = $hostinfo ?
                           $hostinfo->name : $client_sock->peerhost; #(V)
        my $msg = 
            sprintf "[Request for connection from %s]\n", $hostname; #(W)
        warn "$msg";                                                 #(X)
        print $client_sock "Welcome to server script $0 " .
                              "running at $ENV{HOSTNAME}\n\n";       #(Y)
        print $client_sock "Enter your cyberspace name:";            #(Z)
        my $client_name = <$client_sock>;                            #(a)
        $client_name =~ s/^\s*(\S.*\S)\s*\n?$/$1/;                   #(b)
        print $client_sock "Hello $client_name\n\n";                 #(c)
        while ( defined( my $mesg = <$client_sock> ) ) {             #(d)
            print "String receieved from Client: $mesg\n";
            print $client_sock "$client_name: $mesg\n";              #(e)
        }                                                            #(f)
        $client_sock->close or warn $@;                              #(g)
        $msg = sprintf "[Connection closed by %s from %s]\n", 
                       $client_name, $hostname;                      #(h)
        warn "$msg";                                                 #(i)
        exit( 0 );                                                   #(j)
    }
}                                                        
close $server_sock;                                                  #(k)

File: projects/demos/engineering.purdue.edu/scriptingwithobjects/swocode/chap15/AcceptAndForkServer.py
#!/usr/bin/env python

###  AcceptAndForkServer.py

import sys, socket, time, os, signal                                 #(A)

def prefix( message ):                                               #(B)
    now = time.ctime()                                               #(C)
    return "[%s] [%s] %s" % (now, os.getpid(), message)              #(D)

def sigchld_sig_handler( signum, frame ):                            #(E)
    os.wait()                                                        

def interrupt_sig_handler( signum, frame ):                          #(F)
    os.kill( os.getpid(), signal.SIGKILL )                           
    
signal.signal( signal.SIGINT,  interrupt_sig_handler )               #(G)
signal.signal( signal.SIGCHLD, sigchld_sig_handler )                 #(H)

client_sock = client_name = client_hostname = None                   #(I)

# Read a line of text from the client side.  Return an empty
# string when EOF is detected.
def get_client_line(client_sock):                                    #(J)
    client_line = ''                                                 
    while 1:                                                         
        client_byte = client_sock.recv(1)                            
        if not client_byte:                                        
            client_line = ''                                                
            break
        elif client_byte == '\n' or client_byte == '\r':               
            client_line += client_byte                               
            break                                                    
        else:                                                        
            client_line += client_byte                               
    return client_line                                               

port = 9000                                                          #(K)

try:                                                                 #(L)
    server_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)  #(M)   
    server_sock.bind( ('', port) )                                   #(N) 
    server_sock.listen(5)                                            #(O)
except socket.error, (value, message):                               #(P)
    if server_sock:                                                  #(Q)
        server_sock.close()                                          #(R)
    print "No server socket: %s" % prefix(message)                   #(S)
    sys.exit(1)                                                      #(T)

print "[Server %s accepting clients]" % sys.argv[0]                  #(U)

while 1:                                                             #(V)
    client_sock = None
    try:                                                             #(W)
        client_sock, address = server_sock.accept()                  #(X)
    except socket.error, (value, message): pass                      #(Y)
    if not client_sock: continue                                     #(Z)
    child_pid = os.fork()                                            #(a)
    if child_pid == 0:                                               #(b)
        if not client_sock:                                          #(c)
            os.kill( os.getpid(), signal.SIGKILL )                   #(d)
            continue                                                 #(e)
        client_hostname, client_side_port = client_sock.getpeername()#(f)
        message = "[Request for connection from %s]\n" \
                                          % client_hostname          #(g)
        print prefix( message )                                      #(h)
        client_sock.send(
          "Welcome to server script %s running at %s" \
                      % (sys.argv[0], os.environ['HOST']) )          #(i)
        client_sock.send( "\nEnter your cyberspace name: " )         #(j)
        # get client name and get rid of the newline terminator:
        client_name = get_client_line( client_sock )[0:-1]           #(k)
        client_sock.send( "Hello %s\n\n" % client_name )             #(l)
        while 1:                                                     #(m)
            message = get_client_line( client_sock )                 #(n)
            if not message:                                          #(o)
                closing_msg = "[Connection closed by %s from %s]\n" \
                          % (client_name, client_hostname)           #(p)
                print prefix(closing_msg)                            #(q)
                client_sock.close()                                  #(r)
                break                                                #(s)
            else:                                                    #(t)
                # get rid of the newline terminator
                message = message[0:-1]                              #(u)
                client_sock.send("%s: %s\n" % (client_name, message))#(v)
        os.kill( os.getpid(), signal.SIGKILL )                       #(w)
    
server_sock.close()                                                  #(x)
